### DISCUSSION TITLE
ibpython3.7m.dylib can't be opened because Apple cannot check it for malicious software

### DISCUSSION THREAD

**Discussion by ryannfinn** [pyinstaller/pyinstaller#6238] [Help]
I am trying to build a Pyinstaller executable (an electron-eel one) for MacOS, under a Rosetta environment with my M1 MacBook. If I simply pass the Pyinstaller options with --onefolder and without specifying the code-signing option, I can run the program on my own machine after patching some of the libraries like libavcodec on my own. But I cannot run the program on machines other than mine. I have tried this on a few different MacBook models, both M1 and Intel-64 build, but still the following error:

> libpython3.7m.dylib can't be opened because Apple cannot check it for malicious software
code signature not valid for use in process using Library Validation: library load disallowed by system policy

I have then tried to package the exe with code-signing flag with a self-signed certificate, but with the --onefile option, the follow error comes up when I opened the executable:

> not valid for use in process using Library Validation: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)

self-signing with --onefolder option gives this error:

> SystemError: codesign failure!

I then tried to make the exe with --onefile option with only ad-hoc signing (no --code-signing flag passed to PyInstaller), then this is solved but again the libavcodec libraries error come up again. Since I have to patch the libraries manually for them to work, the --onefile option is not an option for me. Basically I have run out of ways to solve the issue here. Can anyone tell me what I should do to handle these conflicts? Thanks!

**Comment by rokm** [comment]
Hmm... so one way I am able to trigger the:

```
[3338] Error loading Python lib '[...]/program/Python': dlopen: dlopen([...]/program/Python, 10): no suitable image found.  Did find:
	[...]/program/Python: code signature in ([...]/program/Python) not valid for use in process using Library Validation: library load disallowed by system policy
```
is if I transfer the onedir bundle between two systems in such a way that the target OS will tag it with `com.apple.quarantine`. For example, build the program on system A, zip the bundle, upload the zip to my web server, then download it on system B with Safari, and then unpack it there. This results in *all* bundle files being tag with `quarantine` attribute:
```
$ xattr program/
com.apple.quarantine
$ xattr program/program 
com.apple.quarantine
$ xattr program/Python 
com.apple.quarantine
```
Since executable is quarantined, it fails to start at all, with UI dialog: `"program" can't be opened because Apple cannot check it for mailicious software." (and I get `Killed: 9` in the terminal).

If I remove the attribute from executable:
```
$ xattr -d com.apple.quarantine program/program
```
and retry running, I get the UI dialog again, but this time "Python" (the shared library, which in your case is called `libpython3.7m`) cannot be opened because Apple cannot check if for malicious software. And in terminal, I get:
```
$ ./program/program 
[3469] Error loading Python lib '[...]/program/Python': dlopen: dlopen([...]/program/Python, 10): no suitable image found.  Did find:
	[...]/program/Python: code signature in ([...]/program/Python) not valid for use in process using Library Validation: library load disallowed by system policy
```

And if I recursively remove the quarantine flags from all files in the bundle:
```
$ xattr -r -d com.apple.quarantine program/
```

The program starts as expected.

So can you check if your bundle is also tagged with quarantine? Did you remove the quarantine from the executable but not from other files in the bundle? And, can you try removing the quarantine flag from the whole bundle?

**Comment by rokm** [comment]
> I have then tried to package the exe with code-signing flag with a self-signed certificate, but with the --onefile option, the follow error comes up when I opened the executable:
> 
> > not valid for use in process using Library Validation: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)

Yes, this is expected, because currently if PyInstaller is given a codesign identity, it also turns on hardened runtime, which does not seem to work with self-signed certificates:
https://github.com/pyinstaller/pyinstaller/blob/632ad00ae3dde2e84953978a59c1b62c98f1ec46/PyInstaller/utils/osx.py#L329-L332

Can you try patching your pyinstaller installation and remove the enabling of hardened runtime, then try with self-signed certificates again? If that fixes your original problem, then we should probably make hardened runtime a separate option.

**Comment by rokm** [comment]
> self-signing with --onefolder option gives this error:
> 
> > SystemError: codesign failure!

What is the preceding output when you get get that error? Does that go away if you use tip of the `v4` branch (or `develop`), which has #6188?

**Comment by rokm** [comment]
> I then tried to make the exe with --onefile option with only ad-hoc signing (no --code-signing flag passed to PyInstaller), then this is solved but again the libavcodec libraries error come up again. Since I have to patch the libraries manually for them to work, the --onefile option is not an option for me.

And lastly, what errors are you running in with libav/ffmpeg libraries?

**Reply by ryannfinn** [reply]
> Reason: Incompatible library version: libavcodec.58.134.100.dylib requires version 9.0.0 or later, but libwebp.7.dylib provides version 8.0.0

The way I solved it is to copy the files of libjpeg.9.dylib and libwebp.7.dylib from /usr/local/Cellar/ after brew installing them in my machine and paste them onto /dist/app.

**Comment by ryannfinn** [comment]
I tried to patch the --runtime flag on pyinstaller/Pyinstaller/utils/osx.py and patched the libav/ffmpeg libraries, and it works on my M1 MacBook and another M1 MacBook. However, as I did that once again for another branch of my software, it didn't work. The code sign failure came up again. Then I used the develop branch with patching to delete the --run-time flag, with patching the libav/ffmepg libraries, the app can be opened on my MacBook again. Nevertheless, on other MacBooks (including intel-64 and M1), the app can only be successfully running after executing with `$ xattr -r -d com.apple.quarantine program/` command. I don't know what exactly is happening here, but it works at the end. Thank you so much for helping!
